### ISSUE TITLE
IO: Node street count float in string results 

### ISSUE THREAD

**Issue by EwoutH** [gboeing/osmnx#727]
**Problem description**
Using [osmnx.io.load_graphml](https://osmnx.readthedocs.io/en/stable/osmnx.html#osmnx.io.load_graphml) to import a GraphML file (saved with `io.save_graphml`) resulted in an error. I traced it back to the `street_count` being saved as an float number inside a string (so `'3.0'`), which errored when `io.load_graphml` tried to convert it directly to int.

**Environment information**
Python 3.7.5, OSMnx 1.1.1.

**Provide a complete minimal reproducible example**
Try to load the `road-graph-def.graphml` file (included in the attached ZIP file below) with `io.load_graphml`:
```python
`ox.io.load_graphml('road-graph-def.graphml')`
```
Output:
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-15c0ad97652b> in <module>()
      1 # Load the MultiDiGraph and convert to DiGraph
----> 2 G2b = ox.io.load_graphml('road-graph-def.graphml', node_dtypes=None, edge_dtypes=None, graph_dtypes=None)
      3 G2a = ox.utils_graph.get_digraph(G2b,weight='free_flow_time')

1 frames
/usr/local/lib/python3.7/dist-packages/osmnx/io.py in _convert_node_attr_types(G, dtypes)
    290     for _, data in G.nodes(data=True):
    291         for attr in data.keys() & dtypes.keys():
--> 292             data[attr] = dtypes[attr](data[attr])
    293     return G
    294 

ValueError: invalid literal for int() with base 10: '3.0'
```
**Mitigation**
I resolved the error by manually converting al N/A values to a float (`0.0`), and saving the floats as ints.
```python
nodes['street_count'] = nodes['street_count'].fillna(0.0)
nodes['street_count'] = [int(i) for i in nodes['street_count']]
```
This graph now loads fine. It's also included in the ZIP file as `road-graph-test.graphml`

Files to reproduce: [road-graphs.zip](https://github.com/gboeing/osmnx/files/6700488/road-graphs.zip)

**Comment by EwoutH** [comment]
I encountered a few more cases where the string conversion of the `io.save_graphml` function caused trouble. For example converting to DiGraph and running Dijkstra's with NetworkX, it wasn't able to compare weights anymore because they were strings.

When I ran the exact same code without first saving to a GraphML file and loading it back, that wasn't a problem.
```Python
# Calculate all the shortests paths between nodes with Dijkstra's algorithm
paths = dict(nx.all_pairs_dijkstra_path(G2a, cutoff=None, weight="free_flow_time"))
```

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-17-81c5b01689fa> in <module>()
      1 # Calculate all the shortests paths between nodes with Dijkstra's algorithm
----> 2 paths = dict(nx.all_pairs_dijkstra_path(G2a, cutoff=None, weight="free_flow_time"))

4 frames
/usr/local/lib/python3.7/dist-packages/networkx/algorithms/shortest_paths/weighted.py in _dijkstra_multisource(G, sources, weight, pred, paths, cutoff, target)
    822             if cost is None:
    823                 continue
--> 824             vu_dist = dist[v] + cost
    825             if cutoff is not None:
    826                 if vu_dist > cutoff:

TypeError: unsupported operand type(s) for +: 'int' and 'str'
```

**Comment by gboeing** [comment]
Can you provide the full code to reproduce your GraphML file? I'm trying to figure out why `street_count` got saved as a float in the first place (why there were nulls).

**Comment by EwoutH** [comment]
> Can you provide the full code to reproduce your GraphML file?

It's a bit messy, but here's the [full Jupyter notebook](https://colab.research.google.com/drive/1DE6MA7E2gOwGp7m35fzvrv5HQgdS3UQT?usp=sharing) (also available [on GitHub](https://github.com/EwoutH/Helsinki-flood-BEP-project/blob/main/road-graph/OSMnx_road_graph_compilation_BEP.ipynb)).

> I'm trying to figure out why `street_count` got saved as a float in the first place (why there were nulls).

Could it have something to do with the simplification of the graph?

> FWIW, if you edited your graph in some way prior to saving it, the `node_dtypes` argument exists to allow you to handle unexpected data, such as nulls in an integer attribute: `ox.io.load_graphml('road-graph-def.graphml', node_dtypes={'street_count': float})`

Thanks!

**Comment by gboeing** [comment]
Can you share a *minimal* reproducible example to produce the ValueError you're getting? If we can narrow this to a minimal code snippet, it may be a bug worth fixing. But as of now it seems like the kind of use case that the `node_dtypes` argument exists to serve: if graph node/edge attributes end up with unexpected or nonstandard data types (such as nulls in `street_count`, when by default there wouldn't be), the user can just pass the `node_dtypes` argument to handle those attributes upon reloading the GraphML file.

**Comment by EwoutH** [comment]
> Can you share a _minimal_ reproducible example

Yes I will. I'm currently finishing my paper, after that I will try to distil it. Should be in an week or two.

**Comment by christophsteinacker** [comment]
I have the same issue and I've been able to track it down to the following for my use case:

The data type of the values for `street_count` changes after consolidating intersections and **then** projecting the graph again.
In the second projection it is irrelevant if I use epsg:4326 or epsg:2955 for `to_crs`.

```python
import osmnx as ox

def print_dtype(g):
    print(set([type(g.nodes[n]['street_count']) for n in g.nodes if 'street_count' in g.nodes[n].keys()]))

# After downloading street_count has dtype int
G = ox.graph_from_place("Piedmont, California, USA", network_type="drive")
print_dtype(G)

# After projection street_count has dtype numpy.int64 but still fine
H = ox.project_graph(G, to_crs='epsg:2955')
print_dtype(H)

# After intersection consolidation street_count still has dtype numpy.int64
I = ox.consolidate_intersections(H, tolerance=10, rebuild_graph=True, dead_ends=True, reconnect_edges=True)
print_dtype(I)

# Projecting the graph changes dtype of street_count to numpy.float64
K = ox.project_graph(I, to_crs='epsg:4326')
print_dtype(K)
```

I hope this helps.

**Comment by gboeing** [comment]
@christophsteinacker I think I understand what's happening in your use case. The `consolidate_intersections` creates some new nodes in the graph to replace old nodes, but does not calculate `street_count` per node for those new ones. This changes the dtype to float so that it can represent the null values. I'll work on a fix.

**Comment by gboeing** [comment]
@christophsteinacker @EwoutH see proposed fix in #782
